plugins/thunderbolt/fu-thunderbolt-device.c

```diff
@@ -320,6 +320,71 @@ fu_thunderbolt_device_get_attr_uint16(FuThunderboltDevice *self, const gchar *na
 	return (guint16)val;
 }
 
+static gboolean
+fu_thunderbolt_device_set_port_offline(FuUdevDevice *device)
+{
+	g_autoptr(GError) error_local_offline = NULL;
+	g_autoptr(GError) error_local_rescan = NULL;
+
+	if (!fu_udev_device_write_sysfs(device, "usb4_port1/offline", "1", &error_local_offline)) {
+		g_warning("Setting port offline failed: %s", error_local_offline->message);
+		return FALSE;
+	}
+	if (!fu_udev_device_write_sysfs(device, "usb4_port1/rescan", "1", &error_local_rescan)) {
+		g_warning("Rescan on port failed: %s", error_local_rescan->message);
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static void
+fu_thunderbolt_device_set_port_online(FuUdevDevice *device)
+{
+	FuUdevDevice *udev = FU_UDEV_DEVICE(device);
+	g_autoptr(GError) error_local = NULL;
+
+	if (!fu_udev_device_write_sysfs(udev, "usb4_port1/offline", "0", &error_local)) {
+		g_warning("Setting port online failed: %s", error_local->message);
+	}
+}
+
+gboolean
+fu_thunderbolt_device_open(FuDevice *device, GError **error)
+{
+	FuThunderboltDevice *self = FU_THUNDERBOLT_DEVICE(device);
+	GUdevDevice *udev_device = NULL;
+	g_autoptr(GUdevDevice) udev_parent = NULL;
+	g_autoptr(FuUdevDevice) parent = NULL;
+	if (self->device_type != FU_THUNDERBOLT_DEVICE_TYPE_RETIMER ||
+	    fu_thunderbolt_device_get_version(self, NULL))
+		return TRUE;
+
+	udev_device = fu_udev_device_get_dev(FU_UDEV_DEVICE(device));
+	udev_parent = g_udev_device_get_parent(udev_device);
+	udev_parent = g_udev_device_get_parent(udev_parent);
+	parent = fu_udev_device_new(g_steal_pointer(&udev_parent));
+	return fu_thunderbolt_device_set_port_offline(parent);
+}
+
+gboolean
+fu_thunderbolt_device_close(FuDevice *device, GError **error)
+{
+	FuThunderboltDevice *self = FU_THUNDERBOLT_DEVICE(device);
+	GUdevDevice *udev_device = NULL;
+	g_autoptr(GUdevDevice) udev_parent = NULL;
+	g_autoptr(FuUdevDevice) parent = NULL;
+	if (self->device_type != FU_THUNDERBOLT_DEVICE_TYPE_RETIMER)
+		return TRUE;
+
+	udev_device = fu_udev_device_get_dev(FU_UDEV_DEVICE(device));
+	udev_parent = g_udev_device_get_parent(udev_device);
+	udev_parent = g_udev_device_get_parent(udev_parent);
```

Comment:
this still leaks

Reply:
let me check on this

Reply:
this should fix https://github.com/fwupd/fwupd/pull/4044/commits/28e70548b8869cbb3ab26804d4770bab43a27718 ?